chef/cookbooks/utils/libraries/role_recipe.rb

```diff
@@ -14,9 +14,44 @@
 #
 
 module CrowbarRoleRecipe
+  # Blacklist mechanism which uses role recipes to allow skipping of
+  # recipes in certain situations.
+  def self.skip_role(node, barclamp, role)
+    # If we're applying the nova-ha-compute batch, skip other roles to avoid
+    # chef-client drift between nodes (since sync marks are only reset before
+    # the first batch) - https://bugzilla.suse.com/show_bug.cgi?id=1004758
+    if node["crowbar"]["applying_for"].key?("nova") &&
+        node["crowbar"]["applying_for"]["nova"].include?("nova-ha-compute") &&
+        role != "nova-ha-compute"
+      return "not required in nova-ha-compute batch"
+    end
+
+    nil
+  end
+
   # Helper to decide whether a role should be run for a node, depending on the
   # state of the node.
   def self.node_state_valid_for_role?(node, barclamp, role)
+    if node[:state] == "applying"
+      # We always want deployer-client, both for heartbeat but also because it
+      # sets up the ability to use the barclamp library
+      return true if role == "deployer-client"
+
+      skip_reason = skip_role(node, barclamp, role)
+      if skip_reason
+        roles = node["crowbar"]["applying_for"].collect { |k, v| v }.flatten.sort.uniq
+        if roles.include? role
+          raise "BUG: skip_role tried to skip #{role} whilst applying " \
+                "batch #{roles} for #{barclamp}"
+        end
+        Chef::Log.info("Skipping role #{role} whilst applying " \
+                       "batch #{roles} for #{barclamp}: #{skip_reason}")
+        return false
+      end
+
+      return true
```

Comment:
We mumbled and I finally understand this - "passing it should not mean that all other checks can be ignored" refers to other checks which don't exist yet, but which could live further down in this method (i.e. not even including the ones which may be added to `#skip_role` in the future).